tools/github-issues/Azure.Sdk.Tools.GitHubIssues/Reports/BaseReport.cs

```diff
@@ -15,48 +16,80 @@ namespace Azure.Sdk.Tools.GitHubIssues.Reports
     public abstract class BaseReport
     {
         protected IConfigurationService ConfigurationService { get; private set; }
+        protected ILogger<BaseReport> Logger { get; private set; }
 
-        public BaseReport(IConfigurationService configurationService)
+        public BaseReport(IConfigurationService configurationService, ILogger<BaseReport> logger)
         {
             ConfigurationService = configurationService;
+            Logger = logger;
         }
 
-        public async Task ExecuteAsync(ILogger log)
+        protected async Task ExecuteWithRetryAsync(int retryCount, Func<Task> action)
         {
-            try
-            {
-                Task<string> pendingGitHubPersonalAccessToken = ConfigurationService.GetGitHubPersonalAccessTokenAsync();
-
-                var gitHubClient = new GitHubClient(new ProductHeaderValue("github-issues"))
+            await Policy
+                .Handle<AbuseException>()
+                .Or<RateLimitExceededException>()
+                .RetryAsync(10, async (ex, retryCount) =>
                 {
-                    Credentials = new Credentials(await pendingGitHubPersonalAccessToken)
-                };
+                    TimeSpan retryDelay = TimeSpan.FromSeconds(10); // Default.
+
+                    switch (ex)
+                    {
+                        case AbuseException abuseException:
+                            retryDelay = TimeSpan.FromSeconds((double)abuseException.RetryAfterSeconds);
+                            Logger.LogWarning("Abuse exception detected. Retry after seconds is: {retrySeconds}",
+                                abuseException.RetryAfterSeconds
+                                );
+                            break;
 
-                log.LogInformation("Preparing report execution context for: {reportType}", this.GetType());
+                        case RateLimitExceededException rateLimitExceededException:
+                            retryDelay = rateLimitExceededException.GetRetryAfterTimeSpan();
+                            Logger.LogWarning(
+                                "Rate limit exception detected. Limit is: {limit}, reset is: {reset}, retry seconds is: {retrySeconds}",
+                                rateLimitExceededException.Limit,
+                                rateLimitExceededException.Reset,
+                                retryDelay.TotalSeconds
+                                );
+                            break;
+                    }
 
-                var context = new ReportExecutionContext(
-                    log,
-                    await ConfigurationService.GetFromAddressAsync(),
-                    await ConfigurationService.GetSendGridTokenAsync(),
-                    await pendingGitHubPersonalAccessToken,
-                    await ConfigurationService.GetRepositoryConfigurationsAsync(),
-                    gitHubClient
-                    );
+                    await Task.Delay(retryDelay);
```

Comment:
Log the fall-through case too; I know we shouldn't get here, but if another exception gets added to the Policy handle we'll want it as a default.

Reply:
Yeah, good point, I'll apply the 10 second backoff (maybe with some jitter). It should never happen though.